src/common/oclengine.cpp

```diff
@@ -52,12 +77,97 @@ void OCLEngine::SetDeviceContextPtrVector(std::vector<DeviceContextPtr> vec, Dev
 
 void OCLEngine::SetDefaultDeviceContext(DeviceContextPtr dcp) { default_device_context = dcp; }
 
-OCLEngine::OCLEngine() { InitOCL(); }
+OCLEngine::OCLEngine()
+{
+    OCLInitResult res = InitOCL(false);
+    all_device_contexts = res.all_device_contexts;
+    default_device_context = res.default_device_context;
+}
 OCLEngine::OCLEngine(OCLEngine const&) {}
 OCLEngine& OCLEngine::operator=(OCLEngine const& rhs) { return *this; }
 
-void OCLEngine::InitOCL()
+cl::Program OCLEngine::MakeProgram(
+    bool buildFromSource, cl::Program::Sources sources, std::string path, std::shared_ptr<OCLDeviceContext> devCntxt)
+{
+    FILE* clBinFile;
+    cl::Program program;
+    cl_int buildError = -1;
+    std::vector<int> binaryStatus;
+    if (!buildFromSource && (clBinFile = fopen(path.c_str(), "r"))) {
+        long lSize;
+
+        fseek(clBinFile, 0L, SEEK_END);
+        lSize = ftell(clBinFile);
+        rewind(clBinFile);
+
+        std::vector<unsigned char> buffer(lSize);
+        lSize = fread(&buffer[0], sizeof(unsigned char), lSize, clBinFile);
```

Comment:
Make sure that the amount of data read is equal to the expected number of bytes.

Reply:
Done.